src/wallet/wallet2.cpp

```diff
@@ -9336,47 +9333,55 @@ std::vector<size_t> wallet2::pick_preferred_rct_inputs(uint64_t needed_money, ui
+  // This vector will hold only indices of spendable transfers to speed up double output checks
+  std::vector<size_t> spendable_transfer_indices;
+  spendable_transfer_indices.reserve(m_transfers.size());
+
+  // try to find a single rct input of enough size or collect otherwise spendable transfers
   for (size_t i = 0; i < m_transfers.size(); ++i)
   {
     const transfer_details& td = m_transfers[i];
-    if (!is_spent(td, false) && !td.m_frozen && td.is_rct() && td.amount() >= needed_money && is_transfer_unlocked(td) && td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
-    {
-      if (td.amount() > m_ignore_outputs_above || td.amount() < m_ignore_outputs_below)
+    if
+    (
+      !is_spent(td, false) &&
+      !td.m_frozen &&
+      td.is_rct() &&
+      is_transfer_unlocked(td) &&
+      td.m_subaddr_index.major == subaddr_account &&
+      subaddr_indices.count(td.m_subaddr_index.minor) == 1 &&
+      !td.m_key_image_partial &&
+      td.amount() <= m_ignore_outputs_above &&
+      td.amount() >= m_ignore_outputs_below
+    )
+    {
+      if (td.amount() >= needed_money)
+      {
+        LOG_PRINT_L2("We can use " << i << " alone: " << print_money(td.amount()));
+        picks.push_back(i);
+        return picks;
+      }
+      else
       {
-        MDEBUG("Ignoring output " << i << " of amount " << print_money(td.amount()) << " which is outside prescribed range [" << print_money(m_ignore_outputs_below) << ", " << print_money(m_ignore_outputs_above) << "]");
-        continue;
+        // Amount is not enough on its own, but it may be enough in addition to another
+        spendable_transfer_indices.push_back(i);
       }
-      LOG_PRINT_L2("We can use " << i << " alone: " << print_money(td.amount()));
-      picks.push_back(i);
-      return picks;
     }
   }
 
   // then try to find two outputs
   // this could be made better by picking one of the outputs to be a small one, since those
   // are less useful since often below the needed money, so if one can be used in a pair,
   // it gets rid of it for the future
-  for (size_t i = 0; i < m_transfers.size(); ++i)
+  for (size_t si = 0; si < spendable_transfer_indices.size(); ++si)
   {
-    const transfer_details& td = m_transfers[i];
-    if (!is_spent(td, false) && !td.m_frozen && !td.m_key_image_partial && td.is_rct() && is_transfer_unlocked(td) && td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
-    {
-      if (td.amount() > m_ignore_outputs_above || td.amount() < m_ignore_outputs_below)
-      {
-        MDEBUG("Ignoring output " << i << " of amount " << print_money(td.amount()) << " which is outside prescribed range [" << print_money(m_ignore_outputs_below) << ", " << print_money(m_ignore_outputs_above) << "]");
-        continue;
-      }
+      const size_t i = spendable_transfer_indices[si];
+      const transfer_details& td = m_transfers[i];
```

Comment:
NIT: I always like to use .at() instead of []. It's just a habit I've built because it will notify you via an exception if you accidentally out-of-bounds it. I see a case where someone accidentally does:

```
const size_t i = spendable_transfer_indices[si];
const transfer_details& td = m_transfers[si]; // Forgot to use i
```

And it doesn't get picked up in a code-review.